src/syn/io/der.rs

```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        if let Class::Universal = &class {
+            if let (PC::Primitive, 2) = (&pc, &tag) {} else {
+                return Err(Error::InvalidType)
+            }
+        }
+
+        eprintln!("[number] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Length::Definite(l) = length {
+            self.buffer.read_i64_number(l).map(T::from_i64)
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_utf8string<C: utf8string::Constraint>(&mut self) -> Result<String, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[utf8string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 12 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
+
+        if let Length::Definite(l) = length {
+            let octets = self.buffer.read_octet_string(l)?;
+            String::from_utf8(octets).map_err(|_| Self::Error::InvalidUtf8String)
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_ia5string<C: ia5string::Constraint>(&mut self) -> Result<String, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[ia5string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 22 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
+
+        if let Length::Definite(l) = length {
+            let octets = self.buffer.read_octet_string(l)?;
+            String::from_utf8(octets).map_err(|_| Self::Error::InvalidUtf8String)
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_octet_string<C: octetstring::Constraint>(&mut self) -> Result<Vec<u8>, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[octet string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 4 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
```

Comment:
With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 4) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_OCTET_STRING != tag {
    return Err(Error::InvalidType);
}
```